File: leetCode/backTrack/Problem996.java
package leetCode.backTrack;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Given an array A of non-negative integers, the array is squareful if for every pair of adjacent elements,
 * their sum is a perfect square.
 * Return the number of permutations of A that are squareful.
 * Two permutations A1 and A2 differ if and only if there is some index i such that A1[i] != A2[i].
 */
public class Problem996 {

    public static void main(String[] args) {
        Problem996 p = new Problem996();
        System.out.println(p.numSquarefulPerms(new int[]{1, 17, 8}));
        System.out.println(p.numSquarefulPerms(new int[]{2, 2, 2}));
        System.out.println(p.numSquarefulPerms(new int[]{0, 0, 0, 1, 1, 1}));
    }

    private int res;

    public int numSquarefulPerms(int[] A) {
        res = 0;
        Arrays.sort(A);
        backTrack(A, new ArrayList<>(), new boolean[A.length]);
        return res;
    }

    private void backTrack(int[] A, List<Integer> list, boolean[] used) {
        if (list.size() == A.length) {
            res++;
        } else {
            for (int i = 0; i < A.length; ++i) {
                if (used[i] || i > 0 && A[i] == A[i - 1] && !used[i - 1]) continue;
                if (list.size() > 0) {
                    int sq = (int) Math.sqrt(list.get(list.size() - 1) + A[i]);
                    if (sq * sq != list.get(list.size() - 1) + A[i]) continue;
                }
                used[i] = true;
                list.add(A[i]);
                backTrack(A, list, used);
                list.remove(list.size() - 1);
                used[i] = false;
            }
        }
    }
}

File: leetCode/backTrack/Problem996.cpp
#include <algorithm>
#include <cmath>
#include <vector>

void backTrack(std::vector<int>&, std::vector<int>&, std::vector<bool>&, int&);
int numSquarefulPerms(std::vector<int>& A) {
  int res = 0;
  std::sort(A.begin(), A.end());
  std::vector<int> list;
  std::vector<bool> used(A.size(), false);
  backTrack(A, list, used, res);
  return res;
}

void backTrack(std::vector<int>& A, std::vector<int>& list,
               std::vector<bool>& used, int& res) {
  if (list.size() == A.size()) {
    res++;
    return;
  }
  for (int i = 0; i < A.size(); ++i) {
    if (used[i] || (i > 0 && A[i] == A[i - 1] && !used[i - 1])) continue;
    // here (A[i] == A[i - 1] && !used[i - 1]) is much faster than
    // (A[i] == A[i - 1] && used[i - 1])
    if (list.size() > 0) {
      int sq = (int)std::sqrt(list.back() + A[i]);
      if (sq * sq != list.back() + A[i]) continue;
    }
    used[i] = true;
    list.push_back(A[i]);
    backTrack(A, list, used, res);
    list.pop_back();
    used[i] = false;
  }
}